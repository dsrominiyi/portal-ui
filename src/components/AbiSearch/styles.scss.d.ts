export const hidden: string;
export const searchBar: string;
