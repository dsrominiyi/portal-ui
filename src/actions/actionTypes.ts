export const SEARCH_CUSTOMER_FETCH_HITS = 'SEARCH_CUSTOMER_FETCH_HITS';
export const SEARCH_ABI_FETCH_HITS = 'SEARCH_ABI_FETCH_HITS';
