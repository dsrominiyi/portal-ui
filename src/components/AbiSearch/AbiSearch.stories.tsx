import React from 'react';
import { Provider } from 'react-redux';
import { storiesOf } from '@storybook/react';
import createMockStore from 'redux-mock-store';

import AbiSearch from '.';

const mockStore = createMockStore([]);

const stories = storiesOf('AbiSearch', module);
const initialProps = {
  codeListId: '403',
  onChange: () => null
};

const render = (search: SearchResults, questionId?: string) => () => (
  <Provider store={mockStore({ search, questionId })}>
    <AbiSearch {...initialProps} />
  </Provider>
);

export const storyRenders = {
  searchByDescription: render(
    {
      suggestions: [
        { id: 'abc', score: 0, suggestion: 'Mr' },
        { id: 'def', score: 0, suggestion: 'Mrs' },
        { id: 'hij', score: 0, suggestion: 'Ms' },
        { id: 'klm', score: 0, suggestion: 'Miss' }
      ],
      query: 'mr',
      found: 4
    },
    '403'
  )
};

stories.add('ABI Description as searchterm', storyRenders.searchByDescription);
