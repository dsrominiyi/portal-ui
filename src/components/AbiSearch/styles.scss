@import '../../style/vars';

.hidden {
  display: none;
}

.search-bar {
  input {
    max-width: 448px;
  }
}
