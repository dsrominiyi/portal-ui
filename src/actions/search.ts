import { SEARCH_CUSTOMER_FETCH_HITS, SEARCH_ABI_FETCH_HITS } from './actionTypes';

const searchUrl = {
  customer: 'https://c86flyhe3i.execute-api.eu-west-1.amazonaws.com/dev/search',
  abiCodes: 'https://ssy2s4rqpl.execute-api.eu-west-1.amazonaws.com/dev/search'
};

export const customerSearchFetchResults = (searchTerm: string) => ({
  type: SEARCH_CUSTOMER_FETCH_HITS,
  payload: searchTerm
    ? fetch(`${searchUrl.customer}?q=${searchTerm}`).then((response: Response) => {
        return response.json();
      })
    : undefined
});

export const abiSearchFetchResults = (searchTerm: string, codeListId: string) => ({
  type: SEARCH_ABI_FETCH_HITS,
  payload: searchTerm
    ? fetch(
        `${searchUrl.abiCodes}?q.parser=structured&q=(and+(term+field%3Dcodelist_id+'${codeListId}')(prefix field%3Ddescription '${searchTerm}'))`
      ).then((response: Response) => response.json())
    : undefined
});
