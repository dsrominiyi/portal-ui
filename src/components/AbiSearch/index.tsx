import React, { FunctionComponent, MouseEvent, useState, Fragment } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import SearchBar from '../SearchBar';
import TextField from '../TextField';
import { abiSearchFetchResults as searchFetchResultsAction } from '../../actions/search';

import styles from './styles.scss';
import formStyles from '../../forms/styles.scss';

export interface Props {
  label?: string;
  codeListId: string;
  attributes?: { [attr: string]: string };
  onChange(value: string): void;
}

export const AbiSearch: FunctionComponent<Props> = ({
  label,
  codeListId,
  attributes,
  onChange
}: Props) => {
  const [isResultSelected, setIsResultSelected] = useState(false);
  const [resultText, setResultText] = useState('');

  const dispatch = useDispatch();
  const {
    search: { abiSearchResults }
  } = useSelector((state: CombinedState) => ({
    search: state.search as SearchState
  }));

  const searchFetchResults = (searchTerm: string) => {
    return dispatch(searchFetchResultsAction(searchTerm, codeListId));
  };

  const onClickResult = (event: MouseEvent) => {
    const result = event.currentTarget.textContent || resultText;
    setResultText(result);
    setIsResultSelected(true);
    onChange(result);
  };

  const onFocus = () => {
    setIsResultSelected(false);
  };

  const testId = attributes && attributes['test-id'];

  return (
    <Fragment>
      <TextField
        key="1"
        value={resultText}
        label={label}
        additionalClasses={[formStyles.formTextField, !isResultSelected ? styles.hidden : '']}
        onChange={() => null}
        onFocus={onFocus}
        attributes={{ 'test-id': testId || '' }}
      />
      <SearchBar
        key="0"
        label={label}
        searchResults={abiSearchResults}
        searchFetchResults={searchFetchResults}
        buildResultText={fields => fields.description}
        additionalClasses={[styles.searchBar, isResultSelected ? styles.hidden : '']}
        onClick={onClickResult}
        attributes={{ 'test-id': testId ? `select-${testId}` : '' }}
      />
    </Fragment>
  );
};

AbiSearch.displayName = 'AbiSearch';
export default AbiSearch;
