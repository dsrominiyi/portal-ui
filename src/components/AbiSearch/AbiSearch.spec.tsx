import { mount, ReactWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';

import { storyRenders } from './AbiSearch.stories';
import testStorySnapshots from '../../helpers/test/testStorySnapshots';

import { SEARCH_ABI_FETCH_HITS } from '../../actions/actionTypes';

let component: ReactWrapper;
const event = {
  currentTarget: {
    textContent: 'Test'
  }
};

const clickResult = () => {
  act(() => {
    (component.find('SearchBar').props() as any).onClick(event);
  });
  component.update();
};

describe('AbiSearch', () => {
  beforeEach(() => {
    component = mount(storyRenders.searchByDescription());
  });

  test('component story renders match snapshots', () => {
    testStorySnapshots(storyRenders);
  });

  test('abiSearchFetchResults dispatches the searchFetchResults action', () => {
    const searchBar = component.find('SearchBar');
    const { searchFetchResults } = searchBar.props() as any;
    const action = searchFetchResults();

    expect(action.type).toBe(SEARCH_ABI_FETCH_HITS);
  });

  test('When a result is selected, the SearchBar component is hidden', () => {
    clickResult();
    const searchBarClassNames = (component.find('SearchBar').props() as any).additionalClasses;

    expect(searchBarClassNames).toContain('hidden');
  });

  test('When the textfield is focused, the SearchBar component is visible', () => {
    clickResult();
    act(() => {
      (component
        .find('TextField')
        .at(0)
        .props() as any).onFocus();
    });
    component.update();
    const searchBarClassNames = (component.find('SearchBar').props() as any).additionalClasses;

    expect(searchBarClassNames).not.toContain('hidden');
  });

  test('When a result is selected, the result text appears in the TextField', () => {
    clickResult();

    expect(
      (component
        .find('TextField')
        .at(0)
        .props() as any).value
    ).toBe(event.currentTarget.textContent);
  });

  test("When no value is selected, the result text retains it's previous value", () => {
    clickResult();
    const clickOnNothingEvent = {
      currentTarget: {
        textContent: undefined
      }
    };
    act(() => {
      (component.find('SearchBar').props() as any).onClick(clickOnNothingEvent);
    });
    component.update();

    expect(
      (component
        .find('TextField')
        .at(0)
        .props() as any).value
    ).toBe(event.currentTarget.textContent);
  });

  test('The TextField onChange takes no action', () => {
    expect(
      (component
        .find('TextField')
        .at(0)
        .props() as any).onChange()
    ).toBe(null);
  });
});
