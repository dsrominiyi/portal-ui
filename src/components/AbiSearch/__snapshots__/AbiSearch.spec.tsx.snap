// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AbiSearch component story renders match snapshots: searchByDescription 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <AbiSearch
    codeListId="403"
    onChange={[Function]}
  />
</ContextProvider>
`;
